src/Build/Evaluation/ItemSpec.cs

```diff
@@ -85,6 +86,25 @@ public override bool IsMatch(string itemToMatch)
                 return ReferencedItems.Any(v => v.ItemAsValueFragment.IsMatch(itemToMatch));
             }
 
+            public override bool IsMatchOnMetadata(IItem item, IEnumerable<string> metadata, MatchOnMetadataOptions options)
+            {
+                if (options.Equals(MatchOnMetadataOptions.CaseSensitive))
+                {
+                    return ReferencedItems.Any(referencedItem =>
+                        metadata.All(m => item.GetMetadataValue(m).Equals(referencedItem.Item.GetMetadataValue(m)) && !item.GetMetadataValue(m).Equals(string.Empty)));
+                }
+                else if (options.Equals(MatchOnMetadataOptions.CaseInsensitive))
+                {
+                    return ReferencedItems.Any(referencedItem =>
+                        metadata.All(m => item.GetMetadataValue(m).ToUpper().Equals(referencedItem.Item.GetMetadataValue(m).ToUpper()) && !item.GetMetadataValue(m).Equals(string.Empty)));
```

Comment:
+1 for using `String.Equals(string, StringComparer)` to specify case sensitivity. Especially for the case insensitive branch, it would be faster than transforming each each string to upper (less allocations, breaks early).